File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st=[]
        for i in s:
            if i in "({[":
                st.append(i)
            elif i in ")}]" and st:
                if i==")" and st[-1]=="(":
                    st.pop()
                elif i=="}" and st[-1]=="{":
                    st.pop()
                elif i=="]" and st[-1]=="[":
                    st.pop()
                else:
                    return False
            else:
                return False
        if(len(st)==0):
            return True
        return False

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<String> st = new Stack<>();
       
            for(int i = 0; i < s.length(); i++)
        {
            String target = s.substring(i,i+1);
//            System.out.println(i+" "+st+" "+target);
            if(target.equals("(") || target.equals("[") || target.equals("{"))
            {
                st.add(target);
            }
            else if(st.isEmpty())
            {
                return false; 
            }
            else if(target.equals(")") && st.peek().equals("("))
            {
                st.pop();
            }
            else if(target.equals("}") && st.peek().equals("{"))
            {
                st.pop();
            }
            else if(target.equals("]") && st.peek().equals("["))
            {
                st.pop();
            }
            else
            {
                return false; 
            }
        }
        if(st.isEmpty())
        {
            return true; 
        }
        else
        {return false; 
        }
    }
}